Persistently map GPU cache data

zakorgy opened the pull request:
Fixes #271.
Right now as a first step the GpuCache texture is replaced with a Buffer. Upon updating the GPU cache we create a writer for the Buffer and update the buffer data directly. The next step would be keeping that writer alive, for that we need to figure out how to handle the lifetimes properly.

kvark commented:
> The next step would be keeping that writer alive, for that we need to figure out how to handle the lifetimes properly.

That would be trivial with [raw mapping](https://docs.rs/gfx-hal/0.2.1/gfx_hal/device/trait.Device.html#tymethod.map_memory), which doesn't have lifetimes.

kvark commented on webrender/src/renderer.rs:
```diff
@@ -751,30 +756,57 @@ impl<B: hal::Backend> GpuCacheTexture<B> {
+
+            // Blit the contents of the previous texture, if applicable.
+            if let Some(blit_source) = blit_source {
+                device.blit_renderable_texture(&mut texture, &blit_source);
```
let's just call the copy directly here, no need to call `blit_xxx` just to early out in there

zakorgy replied:
Fixed in https://github.com/szeged/webrender/commit/99969449fecfb4a124cfdcb88684c4bf8f41e4e8

kvark commented on webrender/src/renderer.rs:
```diff
@@ -751,30 +756,57 @@ impl<B: hal::Backend> GpuCacheTexture<B> {
+            // Blit the contents of the previous texture, if applicable.
+            if let Some(blit_source) = blit_source {
+                device.blit_renderable_texture(&mut texture, &blit_source);
+                device.delete_texture(blit_source);
```
we can't delete the source while it may be in use by the previous frames

zakorgy replied:
`delete_texture` retains textures in the case you mentioned, so we will free the underlying image after that frame is rendered.

zakorgy replied:
Renamed it to `retain_cache_buffer` in https://github.com/szeged/webrender/pull/297/commits/99969449fecfb4a124cfdcb88684c4bf8f41e4e8#diff-f5062b694b9fe53fc1757ed483d577d9R770

kvark commented on webrender/src/renderer.rs:
```diff
@@ -751,30 +756,57 @@ impl<B: hal::Backend> GpuCacheTexture<B> {
+                device.delete_texture(blit_source);
+            }
+
+            device.bind_texture(
```
similarly here, let's not fool ourselves and use an appropriate binding method, e.g. `bind_storage_buffer` or something like this

zakorgy replied:
Renamed it to `bind_cache_buffer` https://github.com/szeged/webrender/pull/297/commits/99969449fecfb4a124cfdcb88684c4bf8f41e4e8#diff-f5062b694b9fe53fc1757ed483d577d9R773

kvark commented on webrender/src/renderer.rs:
```diff
@@ -751,30 +756,57 @@ impl<B: hal::Backend> GpuCacheTexture<B> {
+        {
+            let mut texture = device.create_gpu_cache_texture(
+                TextureTarget::Default,
+                ImageFormat::RGBAF32,
```
most parameters here are useless for this method

zakorgy replied:
Removed the useless parameters in https://github.com/szeged/webrender/commit/99969449fecfb4a124cfdcb88684c4bf8f41e4e8#diff-f5062b694b9fe53fc1757ed483d577d9R762

kvark commented on webrender/src/renderer.rs:
```diff
@@ -873,8 +915,27 @@ impl<B: hal::Backend> GpuCacheTexture<B> {
+                            address,
+                        } => {
+                            let address = address.v as usize * MAX_VERTEX_TEXTURE_WIDTH + address.u as usize;
+                            mapped_ranges.push(address as u64 * 4 .. (address + block_count) as u64 * 4);
```
why are we multiplying by `4`? The block size here is 4 floats, which is 16 bytes

zakorgy replied:
That's right, got confused by the `[f32; 4]`, fixed it here: https://github.com/szeged/webrender/pull/297/commits/61242837ca2384876662a54d0bc03f7d9e4727c5#diff-f5062b694b9fe53fc1757ed483d577d9R931 . Also adding a note to myself to use `GpuBlockData` for the cache buffer related things instead of `[f32; 4]`.

kvark commented on webrender/src/renderer.rs:
```diff
@@ -945,6 +1006,11 @@ impl<B: hal::Backend> GpuCacheTexture<B> {
+            #[cfg(not(feature = "gleam"))]
+            GpuCacheBus::PMbuffer { ref mut mapped_ranges } => {
+                device.flush_mapped_ranges(mapped_ranges.drain(..));
+                return 0
```
no need for `return` here

zakorgy replied:
Fixed in https://github.com/szeged/webrender/pull/297/commits/61242837ca2384876662a54d0bc03f7d9e4727c5

zakorgy commented on webrender/src/device/gfx/device.rs:
```diff
@@ -143,6 +143,7 @@ pub struct Device<B: hal::Backend> {
     programs: FastHashMap<ProgramId, Program<B>>,
     shader_modules: FastHashMap<String, (B::ShaderModule, B::ShaderModule)>,
     images: FastHashMap<TextureId, Image<B>>,
+    gpu_cache_buffers: FastHashMap<TextureId, PMBuffer<B>>,
```
We can have maximum `frame_count` number of buffers here (which is 2 or 3 depending on the present mode), if the size changes in every frame, but that could be a rare case, so what you described is correct.

zakorgy commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1882,6 +1890,129 @@ impl<B: hal::Backend> Device<B> {
+                // to 1 it means we can use that type for our buffer. So this code finds the first
+                // memory type that has a `1` (or, is allowed), and is visible to the CPU.
+                buffer_req.type_mask & (1 << id) != 0
+                    && mem_type.properties.contains(Properties::CPU_VISIBLE | Properties::CPU_VISIBLE)
```
Ooops, one of those should be `COHERENT`

zakorgy replied:
Fixed the mask: https://github.com/szeged/webrender/pull/297/commits/61242837ca2384876662a54d0bc03f7d9e4727c5#diff-cdf03c96bd34cb98c2d09b0be7658232R1918

zakorgy commented on webrender/res/gpu_cache_gfx.glsl:
```diff
@@ -0,0 +1,131 @@
+
+#ifdef WR_VERTEX_SHADER
+
+vec4[8] fetch_from_gpu_cache_8(int address) {
```
Added a comment here : https://github.com/szeged/webrender/pull/297/files#diff-46adf8fca809efdc8698e42f69fa2c2bR18

zakorgy commented on webrender/res/ps_split_composite.glsl:
```diff
@@ -16,10 +16,12 @@ struct SplitGeometry {
-    vec4 data1 = TEXEL_FETCH(sGpuCache, uv, 0, ivec2(1, 0));
+    //vec4 data0 = TEXEL_FETCH(sGpuCache, uv, 0, ivec2(0, 0));
+    //vec4 data1 = TEXEL_FETCH(sGpuCache, uv, 0, ivec2(1, 0));
+    vec4 data0 = fetch_from_gpu_cache_1(address);
```
Fixed here : https://github.com/szeged/webrender/pull/297/files#diff-274690354821e38cf2830b826ab3a82bR19

zakorgy commented on webrender/src/device/gfx/buffer.rs:
```diff
@@ -8,10 +8,57 @@ use rendy_memory::{Block, Heaps, MemoryBlock, MemoryUsageValue, Write};
+        slice::from_raw_parts_mut(self.ptr as *mut _, count)
+    }
+
+    pub unsafe fn flush_mapped_ranges(&self, device: &B::Device, ranges: std::vec::Drain<std::ops::Range<u64>>) {
```
Fixed: https://github.com/szeged/webrender/pull/297/commits/61242837ca2384876662a54d0bc03f7d9e4727c5#diff-5dbe2f57bb9273f9a363bcb71ec6834eR35

zakorgy commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1882,6 +1890,129 @@ impl<B: hal::Backend> Device<B> {
+        debug_assert!(self.inside_frame);
+        assert!(!(width == 0 || height == 0 || layer_count == 0));
+
+        if width > self.max_texture_size || height > self.max_texture_size {
```
Removed in https://github.com/szeged/webrender/pull/297/commits/61242837ca2384876662a54d0bc03f7d9e4727c5

zakorgy commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -376,46 +377,70 @@ impl<K, B, F> DescriptorSetHandler<K, B, F>
-                        hal::memory::Dependencies::empty(),
-                        &[barrier],
-                    );
+            if index == 5 {
```
This was removed in https://github.com/szeged/webrender/pull/297/commits/99969449fecfb4a124cfdcb88684c4bf8f41e4e8

zakorgy commented on webrender/src/device/gfx/descriptor.rs:
```diff
@@ -376,46 +377,70 @@ impl<K, B, F> DescriptorSetHandler<K, B, F>
-                        &[barrier],
-                    );
+            if index == 5 {
+                if let Some(barrier) = cache_buffer.transit(hal::buffer::Access::SHADER_READ) {
```
Fixed in https://github.com/szeged/webrender/pull/297/commits/7806537ffc5f9ac48ce19ebaee1d66c70a7fa273

zakorgy commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1280,7 +1282,9 @@ impl<B: hal::Backend> Device<B> {
     }
 
     pub fn reset_state(&mut self) {
+        let gpu_cache_texture = self.bound_textures[5];
```
This was also fixed in https://github.com/szeged/webrender/commit/99969449fecfb4a124cfdcb88684c4bf8f41e4e8

zakorgy commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2020,8 +2152,21 @@ impl<B: hal::Backend> Device<B> {
+        let ref src_buffer = self.gpu_cache_buffers[&src];
+        let ref dst_buffer = self.gpu_cache_buffers[&dst];
+        unsafe {
+            let src_slice = src_buffer.acquire_host_visible_slice(None);
```
Added the buffer to buffer copy in https://github.com/szeged/webrender/pull/297/commits/921f94bcb684c27b80a4ec62b6c18cb1cb3e4458#diff-cdf03c96bd34cb98c2d09b0be7658232R2161

kvark commented on webrender/src/device/gfx/buffer.rs:
```diff
@@ -8,10 +8,61 @@ use rendy_memory::{Block, Heaps, MemoryBlock, MemoryUsageValue, Write};
+}
+
+impl<B: hal::Backend> PMBuffer<B> {
+    pub unsafe fn acquire_host_visible_slice(&self, size: Option<u64>) -> &mut [[f32; 4]] {
```
this should be `&mut self`, since otherwise it allows multiple mutable slices be acquired to the same thing

zakorgy replied:
Fixed in https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-5dbe2f57bb9273f9a363bcb71ec6834eR26

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -1882,6 +1911,114 @@ impl<B: hal::Backend> Device<B> {
+                // to 1 it means we can use that type for our buffer. So this code finds the first
+                // memory type that has a `1` (or, is allowed), and is visible to the CPU.
+                buffer_req.type_mask & (1 << id) != 0
+                    && mem_type.properties.contains(Properties::CPU_VISIBLE | Properties::COHERENT)
```
I think the implementation guarantees that such a memory type exists, i.e. for any usage of a buffer, there has to be a coherent memory type that supports it

zakorgy replied:
That's right `There must be at least one memory type with both the VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT and VK_MEMORY_PROPERTY_HOST_COHERENT_BIT bits set in its propertyFlags`

zakorgy replied:
Fixed it in https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-cdf03c96bd34cb98c2d09b0be7658232R1980

zakorgy replied:
Fixed it in https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-cdf03c96bd34cb98c2d09b0be7658232R1980

kvark commented on webrender/src/renderer.rs:
```diff
@@ -119,7 +119,7 @@ pub const MAX_VERTEX_TEXTURE_WIDTH: usize = 1024;
 /// be resized every frame, which enables GPU debuggers to see if this
 /// is performed correctly.
-const GPU_CACHE_RESIZE_TEST: bool = false;
+const GPU_CACHE_RESIZE_TEST: bool = true;
```
we'll need to flip this back

zakorgy replied:
This was just flipped for the sake of testing

kvark commented on webrender/src/renderer.rs:
```diff
@@ -699,6 +699,11 @@ impl CacheRow {
+    /// Persistently mapped buffer-based updates
+    #[cfg(not(feature = "gleam"))]
+    PMbuffer {
+        mapped_ranges: Vec<std::ops::Range<u64>>,
```
all the mapped ranges tracking could go away if we require the coherency, which is always available for buffers

zakorgy replied:
Removed the ranges in https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-f5062b694b9fe53fc1757ed483d577d9R704

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2020,6 +2158,38 @@ impl<B: hal::Backend> Device<B> {
 
+    pub fn copy_cache_buffer(&mut self, dst: &Texture, src: &Texture) {
+        let ref src_buffer = self.gpu_cache_buffers[&src.id];
+        let ref dst_buffer = self.gpu_cache_buffers[&dst.id];
```
let's assert on both src and dst to be `is_buffer`

zakorgy replied:
Fixed here: https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-cdf03c96bd34cb98c2d09b0be7658232R2188

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2020,6 +2158,38 @@ impl<B: hal::Backend> Device<B> {
+            cmd_buffer.begin();
+
+            let barriers = src_buffer
+                .transit(hal::buffer::Access::TRANSFER_READ)
```
We are missing the `hal::buffer::Access::HOST_WRITE` bit to be transitioned to whenever we write into the buffer. That may explain the glitches you were seeing.

zakorgy replied:
Added it here: https://github.com/szeged/webrender/pull/297/commits/76b255b4b7ec31e659242e04dbd757177777abf4#diff-cdf03c96bd34cb98c2d09b0be7658232R2215

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2076,6 +2184,60 @@ impl<B: hal::Backend> Device<B> {
+                .chain(dst_buffer.transit(hal::buffer::Access::TRANSFER_WRITE));
+
+            cmd_buffer.pipeline_barrier(
+                hal::pso::PipelineStage::FRAGMENT_SHADER | hal::pso::PipelineStage::VERTEX_SHADER
```
we are missing the `HOST` stage here

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2076,6 +2184,60 @@ impl<B: hal::Backend> Device<B> {
+    pub fn prepare_cache_buffer_for_read(&mut self) {
+        unsafe {
+            let ref buffer = &self.gpu_cache_buffers[&self.bound_gpu_cache];
+            if let Some(barrier) = buffer.transit(hal::buffer::Access::SHADER_READ) {
```
I think that's our problem (or at least one of them :) ). GPU cache is supposed to be modified on CPU while it's read on GPU. So we can *not* just have a single `SHADER_READ` access on it here. While the command buffer is executing, our CPU will still want to write some data for the next frames. We basically need to:
  - always use SHADER_READ + HOST_WRITE access on it
  - except for resizes, where we transition it to `TRANSFER_SRC` before we copy it and then throw it out (after the copy is done on GPU)

Oh, wait, I just realized we can't copy it on GPU on resize because we would have to wait (or do something smart) before writing it on CPU afterwards, duh. Let me write something up as a separate comment...

zakorgy commented:
@kvark I've realized the same issue yesterday but couldn't have phrased as good as you did. I was thinking what should we do in the part what you mentioned as `or do something smart` :

zakorgy commented:
@kvark I've realized the same issue yesterday but couldn't have phrased as good as you did. I was thinking what could we do in the part what you mentioned as `or do something smart`, but these solutions uses multiple submits, and don't know how great overhead that would be: 
- Submit the commend buffer after the buffer copy with a fence and a semaphore. We can wait on the fence before the first write on the new buffer and wait on the semaphore on the next submit.
-  The second one would be call submit after each `draw_tile_frame`, and we would call `prepare_gpu_cache` for the next frame (instead of the current one) after the draw, but before the submit. This could make the cache buffer visible for the next gpu cache update, but again we could end up with a lot of submit.

zakorgy commented:
@kvark I've reverted to CPU-side copy on resize, and handling non-coherent GPU cache buffers as well. I assume most of your previous review comments are related to the buffer to buffer copy part, so I will not answer them one by one.
The only thing I'm not sure is how could we use rendy-memory for the new buffery type. I can see that the `Download` `MemoryUsage` would be good for our needs but as I can see the allocator not always chose the best suitable memory type if there isn't enough available space with that type: https://github.com/amethyst/rendy/blob/63e1dc2264808378f9664f999e942689322018f8/memory/src/heaps/mod.rs#L133.
In this case we can't depend on the information that the device has the suitable memory type, we would need some feedback from the allocator that the chosen type is coherent or not. And I can't find such an information

zakorgy commented:
Let's just forget the rendy-memory part, I can see that here is a [properties](https://github.com/amethyst/rendy/blob/release-0.2/memory/src/heaps/mod.rs#L291) function for the allocated memory block, which can give us the required information.

zakorgy commented:
@kvark I've added the code to use `MemoryBlock`from  rendy-memory for the `PMBuffer` but now the rendering is broken:
The examples doesn't draw anything but there is no validation layer error.
When I run wrench I can only see the white window and the following validation layer error: 
```
VUID-VkMappedMemoryRange-size-00685(ERROR / SPEC): msgNum: 0 - vkInvalidateMappedMemoryRanges: Flush/Invalidate size or offset (61145088, 50659328) exceed the Memory Object's upper-bound (327680). The Vulkan spec states: If size is not equal to VK_WHOLE_SIZE, offset and size must specify a range contained within the currently mapped range of memory (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkMappedMemoryRange-size-00685)
    Objects: 1
        [0] 0x1a, type: 8, name: NULL
```

Which is weird because we don't call `vkInvalidateMappedMemoryRanges` directly in our code, and I can't see those ` (61145088, 50659328)` values in the mapped ranges as well.
I've uploaded the code to a different branch: https://github.com/zakorgy/webrender/commit/fd7bf7fddab21498165443d530d641b41437c614

kvark commented:
I commented on the change you linked to. There is at least one error in `flush_xxx` invocation

kvark commented:
Note: when doing transitions of this buffer, we also need to take the proper ranges into account and specify them for the barriers

zakorgy commented:
> I commented on the change you linked to. There is at least one error in `flush_xxx` invocation

Thanks @kvark for pointing me to the wrong offsets, now the redering works, but there are two new validation layer erros, which probably comes from rendy-memory:
```
VUID-VkMappedMemoryRange-size-00685(ERROR / SPEC): msgNum: 0 - vkInvalidateMappedMemoryRanges: Flush/Invalidate size or offset (61145088, 50659328) exceed the Memory Object's upper-bound (62128128). The Vulkan spec states: If size is not equal to VK_WHOLE_SIZE, offset and size must specify a range contained within the currently mapped range of memory (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkMappedMemoryRange-size-00685)
    Objects: 1
        [0] 0x1a, type: 8, name: NULL
```
and 
```
ERROR 2019-07-30T14:47:30Z: gfx_backend_vulkan: 
VALIDATION [UNASSIGNED-CoreValidation-MemTrack-InvalidMap (0)] : VkMapMemory: Attempting to map memory on an already-mapped object 0x1b.
object info: (type: DEVICE_MEMORY, hndl: 27)
```
Probably the second one comes from the fact that we have a bigger chunk allocated which is related to two gpu cache buffers.
But I still don't know where and why do we call `vkInvalidateMappedMemoryRanges`.

kvark commented:
IIRC, rendy-memory does the flushes and invalidations for you if you use its mapping API.

zakorgy commented:
I've checked the above mentioned errors. So the first one:
```
VUID-VkMappedMemoryRange-size-00685(ERROR / SPEC): msgNum: 0 - vkInvalidateMappedMemoryRanges: 
Flush/Invalidate size or offset (61145088, 50659328) exceed the Memory Object's upper-bound (62128128).
The Vulkan spec states: If size is not equal to VK_WHOLE_SIZE,
offset and size must specify a range contained within the currently mapped range of memory
(https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkMappedMemoryRange-size-00685)
    Objects: 1
        [0] 0x1a, type: 8, name: NULL
```
This occurs when we read back the pixels in wrench, and create a reader. I've checked the offset and size values (the error complains about these values) down to the `gfx-backend-vulkan` level and for me it looks like the size value is wrong in the validation layer error, because the value there is equals with offset+size instead of size.

The second one:
```
ERROR 2019-07-30T14:47:30Z: gfx_backend_vulkan: 
VALIDATION [UNASSIGNED-CoreValidation-MemTrack-InvalidMap (0)] : VkMapMemory:
Attempting to map memory on an already-mapped object 0x1b.
object info: (type: DEVICE_MEMORY, hndl: 27)
```
This happens because we map the same memory twice when creating a new storage buffer and the previous mapping is still alive. We can't call unmap memory from the `gfx-hal` API for the old buffer, because that unmaps the entire memory region which can be related to more buffers.

After these I propose not to use rendy-memory for this special use case.
Also on metal we have the following panic with the rendy-memory changes
```
thread 'main' panicked at 'assertion failed: memory_block.properties().contains(hal::memory::Properties::CPU_CACHED)', webrender/src/device/gfx/device.rs:2002:9
```
but it is absent without it.

zakorgy commented:
@kvark it looks like DX12 is not happy with the shader changes in https://github.com/szeged/webrender/pull/297/commits/881f851e28363e3643644c7c9da36a98c8f07726#diff-46adf8fca809efdc8698e42f69fa2c2bR5
```
[11912] D3D12 ERROR: ID3D12Device::CreateGraphicsPipelineState: Root Signature doesn't match Vertex Shader: Shader SRV descriptor range (BaseShaderRegister=5, NumDescriptors=1, RegisterSpace=2) is not fully bound in root signature	
[11912]  [ STATE_CREATION ERROR #688: CREATEGRAPHICSPIPELINESTATE_VS_ROOT_SIGNATURE_MISMATCH]
```
Could be this an issue in gfx-hal ?

kvark commented:
> Could be this an issue in gfx-hal ?

I'd like to see two things first:
  1. what pipeline layout is being used at this point?
  2. what does the generated HLSL look like? [We output it](https://github.com/gfx-rs/gfx/blob/1670bc87317163f96bbaf19d8d4b32553ae38d55/src/backend/dx12/src/device.rs#L443) in `debug` logging level

kvark commented:
> We can't call unmap memory from the gfx-hal API for the old buffer, because that unmaps the entire memory region which can be related to more buffers.

Can you call unmap() from rendy-memory API? I imagine multiple mappings into the same physical memory is something it has to handle.

zakorgy commented:
> 1. what pipeline layout is being used at this point?

PipelineLayout { raw: WeakPtr( ptr: 0x1e8d7cbd000 ), tables: [SRV_CBV_UAV | SAMPLERS, SRV_CBV_UAV | SAMPLERS, SRV_CBV_UAV | SAMPLERS, SRV_CBV_UAV], root_constants: [RootConstant { stages: VERTEX, range: 0..17 }], num_parameter_slots: 8 }

And the [descriptor set layouts](https://gist.github.com/zakorgy/0c90e2e9c15af7d5fcadde81628ad38a) used for the pipeline layout

> 2. what does the generated HLSL look like? [We output it](https://github.com/gfx-rs/gfx/blob/1670bc87317163f96bbaf19d8d4b32553ae38d55/src/backend/dx12/src/device.rs#L443) in `debug` logging level

Created gists for the [vertex](https://gist.github.com/zakorgy/7f34f5cc4cb721501e04431c2ab0f76c) and [fragment](https://gist.github.com/zakorgy/ea3f0a52b7b58aa0ffd80ea1a349de94) HLSL modules

zakorgy commented:
> > We can't call unmap memory from the gfx-hal API for the old buffer, because that unmaps the entire memory region which can be related to more buffers.
> 
> Can you call unmap() from rendy-memory API? I imagine multiple mappings into the same physical memory is something it has to handle.

I didn't find anything like this in rendy-memory. There are two cases for `unmap` functions in rendy-memory:
1. unmapping the entire memory for decicated allocations
2. do nothing with the underlying memory for the rest

kvark commented:
Looks like a problem with shader translation. It should be at set=1 binding=5, but the generated shader puts it in set=2 binding=5.

kvark commented:
cc @grovesNL @msiglreith ^

msiglreith commented:
mmhh I guess this is due to the `readonly` qualifier. Could you check without it? We might not have enough information on the host side to deal with this

kvark commented:
I filed https://github.com/gfx-rs/gfx/issues/2933 with my findings and ideas about a solution to this. Please try @msiglreith 's suggestion in the meantime as a workaround.

zakorgy commented:
@kvark @msiglreith removing the `readonly` qualifier does the trick for d3d12 but gives the following validation layer error with Vulkan:
```
ERROR 2019-08-02T05:38:28Z: gfx_backend_vulkan:
VALIDATION [UNASSIGNED-CoreValidation-Shader-FeatureNotEnabled (0)] : Shader requires vertexPipelineStoresAndAtomics but is not enabled on the device
object info: (type: UNKNOWN, hndl: 0)
```
which doesn't seem to have impact on the rendering result, it's just not good to have.

zakorgy commented:
@kvark I have updated the PR, everything works fine (fixed the validation layer errors with rendy memory and the wrong chosen memory type with Metal from https://github.com/szeged/webrender/pull/297#issuecomment-516774525) with all backends except the above mentioned Vulkan validation layer error.

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2076,6 +2260,25 @@ impl<B: hal::Backend> Device<B> {
+                let dst_slice = slice::from_raw_parts_mut(write_mapping.ptr().as_ptr() as _, read_size as usize);
+                dst_slice.copy_from_slice(src_slice);
+            }
+            unsafe { dst_buffer.flush_mapped_ranges(&self.device, std::iter::once(0..read_size)) };
```
isn't this done by Rendy-memory if we are using its mapping routines?

zakorgy replied:
I found using the `write` and `read` functions form `rendy-memory` here causes segfault (without any backtrace) after a certain time when running wrench tests.

kvark replied:
are you running wrench tests in Debug? please file a bug to rendy, it seems pretty important to investigate

zakorgy replied:
Created an issue : https://github.com/amethyst/rendy/issues/183

zakorgy replied:
It turned out that the issue was on our side see: https://github.com/amethyst/rendy/issues/183#issuecomment-519095649

zakorgy replied:
Removed this in the last commit: https://github.com/szeged/webrender/pull/297/commits/ed9e894391d366dfc69d34e7ef4b33250011e111#diff-cdf03c96bd34cb98c2d09b0be7658232L2275

kvark commented on webrender/src/device/gfx/buffer.rs:
```diff
@@ -4,14 +4,84 @@
+        self.memory_block.unmap(device);
+    }
+
+    pub(super) unsafe fn flush_mapped_ranges(
```
do we need this now as we are using rendy-memory?

zakorgy replied:
I wanted to keep this, since rendy-memory's `MappedRange` has a lifetime, so we can't transfer it if we wan't write the mapped memory in the `RendererBackend` thread.
My idea would be store a pointer in the `RendererBackend` thread, which points to the start of the mapped memory region. We could use this ptr to write the GPUCache data in place and after that we send the ranges to the `Renderer` thread where we call  `flush_mapped_ranges` with those ranges.

kvark replied:
IIRC from our discussion on the call, we don't need to keep the mapped range if we use rendy-memory. We can rely on the fact that internally rendy-memory keeps the mapping open (since it has to). So what we need to do is whenever we need to write something, we ask rendy-memory for the mapped range, write it, and then let it go. There should be no need to keep it around and/or transfer anywhere.

zakorgy replied:
Removed this in the last commit: https://github.com/szeged/webrender/pull/297/commits/ed9e894391d366dfc69d34e7ef4b33250011e111#diff-5dbe2f57bb9273f9a363bcb71ec6834eL36

zakorgy commented:
@kvark I've addressed your last concern (it turned out the issue was on our side, so we are using the mapping routines from rendy-memory), also squashed some commits.
